NIFI-589: Add processors that can run Apache Flume sources/sinks

joey opened the pull request.

busbey commented:
please update commit message to reference the jira. please squish into a single commit.

busbey commented:
could you add a README that explains what the processor set is trying to do and what restrictions it has (e.g. what versions of HDFS does the HDFS sink work with, what versions of the Twitter API does the twitter source work with, etc) or pointers to where people can learn about the components included?

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/NifiChannel.java:
```diff
@@ -0,0 +1,31 @@
+
```
This class needs a license notice at the top.

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/NifiChannel.java:
```diff
@@ -0,0 +1,31 @@
+
+  @Override
+  public void configure(Context context) {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
please leave out the autogenerated comment.

does this need to throw? could it just be a no-op?

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/NifiChannelSelector.java:
```diff
@@ -0,0 +1,55 @@
+
```
this class needs a license header.

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/NifiTransaction.java:
```diff
@@ -0,0 +1,40 @@
+
```
class needs a license header.

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/util/FlowFileEvent.java:
```diff
@@ -0,0 +1,114 @@
+
```
class needs a license header.

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/util/FlowFileEventConstants.java:
```diff
@@ -0,0 +1,25 @@
+
```
class needs a license header

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/test/java/org/apache/nifi/processors/flume/AbstractFlumeTest.java:
```diff
@@ -0,0 +1,35 @@
+    private static Logger logger;
+
+    @BeforeClass
+    public static void setUpClass() {
```
could we move these into a test logger configuration file?

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/test/java/org/apache/nifi/processors/flume/FlumeSinkProcessorTest.java:
```diff
@@ -0,0 +1,165 @@
+            "tier1.sinks.sink-1.hdfs.serializer = TEXT\n" +
+            "tier1.sinks.sink-1.serializer.appendNewline = false"
+        );
+        FileInputStream fis = new FileInputStream("src/test/resources/testdata/records.txt");
```
Please use getResourceAsStream so that this test resource will be loaded out of maven target/. This will allow us to later add pre-processing if necessary.

i.e.

InputStream is = getClass().getResourceAsStream("/testdata/records.txt");

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/test/java/org/apache/nifi/processors/flume/FlumeSinkProcessorTest.java:
```diff
@@ -0,0 +1,165 @@
+    
+    @Test
+    public void testHdfsSink() throws IOException {
+        File destDir = new File("target/hdfs");
```
please get this from junit / maven. For example, by using a TemporaryDirectory JUnit rule.

    @Rule
    public final TemporaryFolder temp = new TemporaryFolder();

    @Test
    public void someTest() {
        File destDir = temp.newFolder("hdfs");

joey replied:
`TemporaryFolder` doesn't use the maven `target` directory. On my Mac, it uses directories that are created in `/var/folders`. Is there any concern that a `mvn clean` doesn't guarantee that those temporary files are cleaned up.

joewitt replied:
mmm...yeah it would be preferable that 'clean' does the job.

busbey replied:
TemporaryFolder guarantees that the folders are destroyed at the end of hte test cycle, so there's nothing for maven clean to handle.

joey replied:
That's not how I read the javadoc:
> The TemporaryFolder Rule allows creation of files and folders that should be deleted when the test method finishes (whether it passes or fails). Whether the deletion is successful or not is not checked by this rule. No exception will be thrown in case the deletion fails.

In particular the phrase "Whether the deletion is successful or not is not checked by this rule." worries me. If you're confident this works as expected I'm ok with it. These tests don't write a lot of data to the filesystem but I worry about cases where a test crashes mid run and could potentially leave a large amount of temporary data on the user's system.

busbey replied:
Yes it works. If you're going to muck with permissions to disallow changes by the creating user (like some HDFS test do), then the deletion attempt will fail and JUnit will not notice. However, in that case maven clean would fail as well.

If you're going to rely on maven clean instead, then you should be getting the location to write to from maven rather than hard-coding 'target' in the working directory.

joey replied:
I think I can have it both ways. I just pushed a change that sets the `java.io.tmpdir` system property for surefire tests to the build project directory.

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/test/java/org/apache/nifi/processors/flume/FlumeSinkProcessorTest.java:
```diff
@@ -0,0 +1,165 @@
+        File[] files = destDir.listFiles((FilenameFilter)HiddenFileFilter.VISIBLE);
+        assertEquals("Unexpected number of destination files.", 1, files.length);
+        File dst = files[0];
+        byte[] expectedMd5 = FileUtils.computeMd5Digest(new File("src/test/resources/testdata/records.txt"));
```
Please use getResourceAsStream so that this test resource will be loaded out of maven target/. This will allow us to later add pre-processing if necessary.

i.e.

InputStream is = getClass().getResourceAsStream("/testdata/records.txt");

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/test/java/org/apache/nifi/processors/flume/FlumeSourceProcessorTest.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Test
+    public void testSourceWithConfig() throws IOException {
+        File spoolDirectory = new File("target/spooldir");
```
please get this from junit / maven. For example, by using a TemporaryDirectory JUnit rule.

    @Rule
    public final TemporaryFolder temp = new TemporaryFolder();

    @Test
    public void someTest() {
        File destDir = temp.newFolder("spooldir");

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/test/java/org/apache/nifi/processors/flume/FlumeSourceProcessorTest.java:
```diff
@@ -0,0 +1,140 @@
+        } else {
+          spoolDirectory.mkdirs();
+        }
+        File src = new File("src/test/resources/testdata/records.txt");
```
Please use getResourceAsStream so that this test resource will be loaded out of maven target/. This will allow us to later add pre-processing if necessary.

i.e.

    File src = new File(getClass().getResource("/testdata/records.txt").toURI());

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/test/resources/core-site.xml:
```diff
@@ -0,0 +1,25 @@
+<configuration>
+    <property>
+        <name>fs.defaultFS</name>
+        <value>file:///</value>
```
could we make this a filtered file so that we can put the defaultFS path into the build directory?

joey replied:
When setting `fs.defaultFS` to `file:///some/path` that doesn't do anything to change the root of the local filesystem that Hadoop classes use, they always start at the root of the local filesystem.

busbey replied:
that is terrible and sounds like a bug.

busbey replied:
ugh. no way to mark concerns "resolved" in github.

so I'll just leave another comment, that I agree the bug in hadoop means there's no sense updating this.

busbey commented on nifi/pom.xml:
```diff
@@ -789,6 +789,11 @@
                 <artifactId>nifi-geo-nar</artifactId>
                 <version>0.1.0-incubating-SNAPSHOT</version>
                 <type>nar</type>
+            <dependency>
```
this change misses closing the geo-nar dependency line, which breaks compilation.

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/pom.xml:
```diff
@@ -0,0 +1,126 @@
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-flowfile-packager</artifactId>
+        </dependency>
+        <dependency> 
```
nit: trailing whitespace

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/pom.xml:
```diff
@@ -0,0 +1,126 @@
+            <artifactId>nifi-flowfile-packager</artifactId>
+        </dependency>
+        <dependency> 
+            <groupId>org.apache.flume</groupId> 
```
nit: trailing whitespace

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/pom.xml:
```diff
@@ -0,0 +1,126 @@
+        </dependency>
+        <dependency> 
+            <groupId>org.apache.flume</groupId> 
+            <artifactId>flume-ng-sdk</artifactId> 
```
nit: trailing whitespace

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/pom.xml:
```diff
@@ -0,0 +1,126 @@
+            <groupId>org.apache.flume</groupId> 
+            <artifactId>flume-ng-sdk</artifactId> 
+            <version>1.5.2</version>
+        </dependency> 
```
nit: trailing whitespace

busbey commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/test/resources/testdata/records.txt:
```diff
@@ -0,0 +1,4 @@
+record 1
```
please either add a license header or add this file to the list of exemptions for apache-rat:check

joewitt commented:
Sean this was a pretty awesome review.  I also really like the Junit temp directory thing.

joewitt commented:
I have reached out to Joey on this.  He plans to incorporate Sean's changes and reach out to the Flume community.  Want to make sure they're comfortable with this as well.  Joey is quite time crunched so if he would like to pass the baton we would be happy to try and pick it up.  Great contrib.

joey commented:
Sorry for the delay, I should have a second iteration of the patch this afternoon PT.

Do you prefer squashing on updates or waiting until the patch is accepted and doing the squash shortly before commit?

joewitt commented:
Squash for the review is easier.  But, seriously, we can deal either way.

joey commented:
Do you want all commits squashed or just the incremental commit? I ask because some people prefer to always review the entire patch as is and others prefer to the changes since the last review. I can easily do either.

joewitt commented:
Ahh i see.  I'd like to see the entire patch as a new entry so it can be reviewed in whole again.

busbey commented:
doing the incremental commit as its own thing means we could do either though. "changed files" tab shows it as the whole and just the commit shows the incremental.

squashing everything means that if things work out we don't have to wait for another round trip of communication for the ready-to-push commit though, and I can use vimdiff with the previous change.

joey commented:
Based on @busbey's last comment, I'll leave out the squash for now in case anyone wants to review just the changes since the last PR.

I'm also going to leave out a rebase unless you want that now. I figured I'd get @busbey's review done and then do a rebase and squash before handing it off to a reviewer from Flume.

joey commented:
This new commit addresses all of the review feedback except the README. I didn't do a rebase, so this patch probably doesn't apply as is. I wanted to post this for feedback separately from the rebase which I'll do next.

I'm not sure about the README. We don't want to get into a situation where we're trying to document all of the features provided by Flume. I think if we add something, it should be a pointer to the Flume documentation for the version of Flume that we're using.

Will that be sufficient?

joey commented:
I did a rebase and a force push. @busbey, when you get a chance can you do a quick review to make sure that I caught all of your feedback?

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSinkProcessor.java:
```diff
@@ -0,0 +1,193 @@
+
+    @OnScheduled
+    public void onScheduled(final SchedulingContext context) {
+        batchSize = context.getProperty(BATCH_SIZE).asInteger();
```
Properties initialized in OnScheduled should be volatile.

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSinkProcessor.java:
```diff
@@ -0,0 +1,193 @@
+    public void onTrigger(final ProcessContext context,
+            final ProcessSession session) throws ProcessException {
+        List<FlowFile> flowFiles = Lists.newArrayListWithExpectedSize(batchSize);
+        for (int i = 0; i < batchSize; i++) {
```
Why not use [`ProcessSession.get(int max)`](https://github.com/apache/incubator-nifi/blob/develop/nifi/nifi-api/src/main/java/org/apache/nifi/processor/ProcessSession.java#L143)?

joey replied:
Because I didn't see that method :)

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSinkProcessor.java:
```diff
@@ -0,0 +1,193 @@
+        try {
+            Sink.Status status;
+            do {
+              status = sink.process();
```
Nit: inconsistent indent

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSourceProcessor.java:
```diff
@@ -0,0 +1,191 @@
+            throws ProcessException {
+        try {
+            pollableSource.setChannelProcessor(new ChannelProcessor(
+                    new NifiChannelSelector(new NifiChannel(session, SUCCESS))));
```
I think the channel, selector, and processor can be reused by setting the session on the channel each time. The current session would be used to create transactions, which we don't expect to be long lived because this is a source. Then you might not have to start and stop the source each time, right?

joey replied:
Yes that should work.

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSourceProcessor.java:
```diff
@@ -0,0 +1,191 @@
+        transaction.begin();
+
+        try {
+            Event event = channel.take();
```
Should this get all of the available events from the channel or at least more than one? The event driven source "provides its own event-driven mechanism to invoke event processing". That sounds like it will produce events on this channel without outside intervention and may have produced many between the time that NiFi triggers the processor. If that's the case, then I think this needs to consume as many as possible.

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSinkProcessor.java:
```diff
@@ -0,0 +1,193 @@
+            flowFiles.add(flowFile);
+        }
+
+        Transaction transaction = channel.getTransaction();
```
Why not wrap the current session in a Flume transaction like the polling source does?

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSourceProcessor.java:
```diff
@@ -0,0 +1,191 @@
+            Configurables.configure(source,
+                    getFlumeSourceContext(flumeConfig, agentName, sourceName));
+
+            if (source instanceof EventDrivenSource) {
```
If I understand correctly, the memory channel is used to buffer incoming events between calls to onTrigger because there is no session with which the events would be sent. But this makes all event-driven sources unsafe because the events that arrive between onTrigger calls are stored in memory.

I think you instead need access to a ProcessSessionFactory. It looks like the Processor interface is passed a session factory when onTrigger is called, and [`AbstractProcessor`](https://github.com/apache/incubator-nifi/blob/develop/nifi/nifi-api/src/main/java/org/apache/nifi/processor/AbstractProcessor.java#L24) is what is giving you a session most of the time. I'm not sure if you can get one otherwise or if you can cache the instance, but I think this should be adding content directly to the repository or using a file channel (which can hopefully be avoided) rather than using a mem channel.

joewitt commented:
Joey - i appreciate you don't have a ton of time to do this and that it is by no means a trivial patch.  I am not in love with the convo flow here in terms of following what is being said or outstanding questions.  If there are discussion items you'd like to go over please do feel free to shoot a conversation started to dev@nifi.iao.  If not needed that is cool too.

joey commented:
Thanks @joewitt.  I've been working through @rdblue's feedback. Right now, I'm seeing if I can have the event driven sources write directly to the NiFi session rather than through a MemoryChannel.

After that, the last bit of feedback for me to take a look at is if I can simplify the logic of the FlumeSinkProcessor to avoid the MemoryChannel and wrap the Transaction directly. I should have an updated patch today or tomorrow.

rdblue commented:
@joey did you find a good way to have a reference to a `ProcessSession` outside of the `onTrigger` method?

joey commented:
@rdblue yes, I need to extend `AbstractSessionFactoryProcessor` instead of `AbstractProcessor`. The current problem that I'm seeing is that my source continues to run after the processor is stopped. I'm looking to see if this is a bug in Flume or a problem with how I'm using the Flume API.

joey commented:
@rdblue and @joewitt See the latest. I think this addresses all of the Ryan's feedback. I do have one open question:

Currently I'm stopping the Sinks and Sources in an `@OnUnscheduled` callback, but I'm not sure if that's the right place to do it. I'm not sure about the intention of the API generally, but the `TestRunner#run()` method will call `@OnUnscheduled` callbacks immediately after the first call to `Processor#onTrigger()`. That means that effectively only a single `onTrigger()` happens in any of my tests. Should this be delayed until an `@OnStopped` callback?

markap14 commented:
@joey generally @OnStopped is the best place to do something like that. If you want to call onTrigger() multiple times without the @OnScheduled, @OnUnscheduled being called, you can pass in flags to the run() method to specify this. Or alternatively, if it makes sense in your unit test, you can use .run(int) to run some specified number of times.

joey commented:
Thanks @markap14!

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSinkProcessor.java:
```diff
@@ -0,0 +1,146 @@
+    public void onScheduled(final SchedulingContext context) {
+        try {
+            channel = new NifiSinkSessionChannel(SUCCESS, FAILURE);
+            Configurables.configure(channel, new Context());
```
This probably isn't needed now that you are using a custom channel implementation.

joey replied:
That's true, I can remove it.

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSinkProcessor.java:
```diff
@@ -0,0 +1,146 @@
+
+        channel.setSession(session);
+        try {
+            if (sink.process() == Sink.Status.BACKOFF) {
```
I don't think backoff should make this call `context.yield()`. In Flume, backoff is used to signal the sink runner that there is no more input, but NiFi already manages things like that. My understanding is that in NiFi, yield means it can't process for some external reason so there is no point in triggering even if there is input. (For example, can't connect to a DB so wait a few seconds for it to come up and retry.) That's not what we want to signal here, though. I think this should just call process and maybe increment a counter for how many times the result is backoff vs ready.

Speaking of counters, is there a way to hijack Flume's counters?

joey replied:
That makes sense. The description of yield wasn't clear to me.

I might be able to get access to Flume counters. Are there docs on the NiFi metrics API?

joewitt replied:
Ryan's view on context.yield is inline with its intent.  Generally it is a way for the processor to communicate to the framework that it doesn't want to be scheduled for a while because some transient issue is occurring that it expects will clear up 'later'.

joewitt replied:
Docs on our metrics API...i believe the metrics information is closed off from processors being able to inject anything custom.  The only game in town * i believe * is the counters mechanism which solely exists for Sean and Tony :-)  Which is admittedly a good enough reason.  What sorts of things would you like to do with the metrics?

joey replied:
It'd be nice if we could expose the counters that Flume uses internally to the NiFi UI. If APIs for that don't exist yet, then there's no reason this feature needs to be blocked by that.

rdblue replied:
Agreed, this doesn't need to be blocked. I also just looked at the [implementation in the DatasetSink](https://github.com/apache/flume/blob/trunk/flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java#L347) and it doesn't look useful anyway. The counters are about the internal batching, which is pretty much all handled by NiFi anyway. And, the sink has to get its own counter object so it looks really hard to forward that information to NiFi's counters.

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSourceProcessor.java:
```diff
@@ -0,0 +1,203 @@
+            if (source instanceof PollableSource) {
+                source.setChannelProcessor(new ChannelProcessor(
+                    new NifiChannelSelector(pollableSourceChannel)));
+                source.start();
```
I don't think this should be started until onTrigger so the ProcessSessionFactory can be set. Otherwise, the channel could start sending events to this source, which will immediately fail because there is no session.

joey replied:
The `PollableSource` will only use the channel on calls to `#process()`

rdblue replied:
Oops! I thought this was like before where the event driven source had extra initialization. Nevermind.

rdblue commented on nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSourceProcessor.java:
```diff
@@ -0,0 +1,203 @@
+            super.onTrigger(context, sessionFactory);
+        } else if (source instanceof EventDrivenSource) {
+            ProcessSessionFactory old = sessionFactoryRef.getAndSet(sessionFactory);
+            if (old == null) {
```
Doesn't this need to happen if `old` is ever not the same as `sessionFactory`? It would be good to log or throw an exception because we don't expect that to ever happen, right?

joey replied:
I based this on how the `ListenUdp` processor is written. I'm happy to throw an exception if it changes out from under us as that should only happen after the processor is stopped and started again and the `@OnStopped` method sets it back to `null`.

rdblue commented:
+1

Looks great!

joewitt commented:
Should we move ahead to Flume 1.6.0 now?  Appears to still compile/test/run fine with that.

nifi/pom.xml  Formatting issue with derby dependency.  Tabs gone wild.  Fix dependency ref to 0.2.0

nifi/nifi-nar-bundles/nifi-flume-bundle/pom.xml  Fix dependency ref to 0.2.0

Two easy check style violations to resolve:

WARNING] src/main/java/org/apache/nifi/processors/flume/FlumeSinkProcessor.java[24:8] (imports) UnusedImports: Unused import - org.apache.flume.Context.
[WARNING] src/main/java/org/apache/nifi/processors/flume/util/FlowFileEvent.java[30] (imports) AvoidStarImport: Using the '.*' form of import should be avoided - org.apache.nifi.processors.flume.util.FlowFileEventConstants.*.

Need to establish a LICENSE and NOTICE within the nifi-flume-nar project so that it will end up that the bar binary will have proper accounting for bundled items and need to roll up any unique dependencies into the nifi-assembly LICENSE and NOTICE.  I reviewed all bundled dependencies and they all appear to be legit so doing this will be tedious but not problematic.  I am happy to do this for you.  There are lots of good examples of how to do this in the other nar bundles and we’ve augmented ASF guidance for handling licensing here; http://nifi.incubator.apache.org/licensing-guide.html

The various Flume processors within the Nar appear to have a significant version conflict for Netty.  Ranging from wanting 3.5.x to 3.8.x of Netty.  Unclear if this is important so just bringing to your attention.

The bundled dependencies include a rather old version of Jetty.  I am not clear if this will conflict with the Jetty libraries we use and a quick scan of our docs suggests we don’t point out the limitations of Nar class loader isolation (will make a ticket).  You may have no action here just bringing it to your attention.

Recommend renaming the processors to ‘ExecuteFlumeSource’ and ‘ExecuteFlumeSink’.  The naming structure we try to follow is generally verb-subject.  I did not find that we had this documented (will make a ticket).

This processor is a perfect case for providing some ‘additional details’ documentation so that when someone looks up the usage it will show them how to configure a fully working example.  http://nifi.incubator.apache.org/docs/nifi-docs/html/developer-guide.html#advanced-documentation

Is there any service provider construct or similar mechanism whereby we could provide a dropdown of identified sources and sinks so the user doesn’t have to type them by hand?  Not critical just curious.  Given that the other property is pasting of a flume properties section this is already pretty ‘developer heavy’ anyway so it isn’t necessary.  But if possible it could be a nice convenience to offer.

FlumeSinkProcessor: I think because channel.setSession(…) is being called as it is during the onTrigger method this means the processor needs to run serially.  If multiple threads are running then the channel’s session would get overwritten potentially creating some very interesting threading issues most likely.  Is marking it is as serial only going to be sufficient?  We have an annotation for that if so which will have the framework ensure it only gets a single thread for the processor.  I think a similar issue is present on FlumeSourceProc.

If you happen to have a template of a flow that can be used to test/demonstrate its function that would be a great thing to have.  If so please add to the JIRA.

Looking forward to seeing this merged in.  Let me know if you’d like to address these or discuss them further.

Thanks
Joe

rdblue commented:
> Is there any service provider construct or similar mechanism whereby we could provide a dropdown of identified sources and sinks

I think this is primarily intended to allow people to use their current Flume config inside of NiFi and pave the way for more seamless integration as a next step. I would say we want to keep this as-is for compatibility (just paste in your current config) and create processors to wrap the widely used Flume components, like the Avro source.

> I think because channel.setSession(…) is being called as it is during the onTrigger method this means the processor needs to run serially

Good point. Maybe we can keep a thread-local cache of the sessions and pass an accessor to the channel instead? Then we would be able to keep parallel execution.

joewitt commented:
> Good point. Maybe we can keep a thread-local cache of the sessions and pass an accessor to the channel instead? Then we would be able to keep parallel execution.

You'll want to find a mechanism other than thread local.  There basically zero guarantees on which thread you will get between 'onTrigger' calls.  Now if the processor takes a thread and doesn't return it to the controller you can but that is pretty naughty.

rdblue commented:
I think thread-local would be okay because the thread doesn't change while the process is running. In onTrigger, you would first set the thread's session, then call the Flume processor that pulls from the channel by accessing the session for its thread.

joewitt commented:
ryan

ok got ya.  I think i'll need to see it once it is implemented to understand it better.  But so long as you understand the tradespace we should be good

Thanks
Joe

joey commented:
I need to look at the threading stuff. The Flume API states that Sink and Source instances are not shared between threads, so we need to maintain that invariant.

joey commented:
Sorry for the delay, I've been looking at the threading model and I'm convinced that for the first version we should mark the processors as serial only. 

This matches the behavior of

joey commented:
Sorry, hit comment by accident. 

This matches the behavior of Flume so there's no loss in functionality for the user. In fact a number of sinks/sources have config parameters to run threads internally for parallelism so users can still take advantage of those. Finally, I reviewed some Flume code and I'm nervous about finding subtle concurrency issues with trying to run multi-threaded sinks/sources. 

If performance is a real concern as users start working with it, we can work on adding it in the future.

I should have an updated PR with the other comments addressed later today. 

As far as a drop down of sources/sinks go, Flume has no service loader mechanism so I can't do it generally. I could populate a combo box with the sources/sinks we ship and users could type in a class if they added one after the fact. 

There is no API for introspecting config parameters so we'll have to keep that a copy/paste experience. I agree with Ryan, if any sources/sinks see a lot of adoption we should re-write them as first-class processors.

joewitt commented:
Joey,

Love it and agree with all the points made.

Ryan, Sean, Mark,

Thanks for the great review effort/input on this.

Thanks
Joe

joey commented:
I rebased and pushed a new commit with most of the review feedback. I'm still working on the extended docs and a template to make testing with a running instance easier. I hope to have those published later today too.

rdblue commented:
+1. Thanks Joey!

rdblue commented:
Another way to fix the threading issue is to validate the sinks and sources are safe and wrap them in first-class processors that allow threading even if the general processor doesn't. The MorphlineSolrSink, for example would run just fine and would be a really valuable processor (I talked with @whoschek about it last week).

Joe, do you think that it would be possible to make the serial-only setting an option that gets exposed to the user?

joey commented:
If there are specific sources/sinks we want to expose as processors that support threading, we could probably do that by adding unique processors for them that wraps the generic version.

joewitt commented:
i'd encourage going with serial only mode for now.  Let's learn more and iterate from there.

joey commented:
Alright, docs pushed. This latest version passes the contrib-check so let me know if you think it needs anything else.

Otherwise, I'll do another rebase and squash the commits into one.

I can also wait to do the rebase and squash until after the 0.2.0 release so that this will be ready to merge after that.

joewitt commented:
Joey,

I just took another visual pass through it all.  I'm not sure what the emoticon for crying tears of joy is but when I saw the license/notice and documentation included...i was basically needing the emoticon.

Will run through a few test scenarios and pull in the code and get back to you.  This week is a little wonky work wise so unless someone else picks up the ball response might be a little slower than this deserves.

Thanks
Joe

joey commented:
You're right, it's a tedious process :) I think the Yetus project (being spun out of Hadoop) is planning to take a stab at having programatic tools for rolling up NOTICES to hopefully make this less tedious.

I didn't roll up the NOTICES/LICENSE file to the nifi-assembly. Let me know if you want me to take a stab at that or if y'all want to handle it.

> The bundled dependencies include a rather old version of Jetty. I am not clear if this will conflict with the Jetty libraries we use and a quick scan of our docs suggests we don’t point out the limitations of Nar class loader isolation (will make a ticket). You may have no action here just bringing it to your attention.

I noticed that too. It's the same version that's in the nifi-hadoop-libraries-nar, so I assume it's ok? I also don't think that any of the Flume components will use Jetty the way we're instantiating them. I think that's used by Flume to publish internal metrics data.

joewitt commented:
NAR documentation fix: https://issues.apache.org/jira/browse/NIFI-763

Tedious licenses: Yeah no worries.  I will roll up it once I merge it in.  Gives me a chance to thoroughly double check stuff anyway.  Thanks.  I feel like we could automate the generation of those project level license/notices anyway (yetus should put a good dent in this though).

thanks

joey commented:
I meant to upload these yesterday, but I uploaded a couple of templates to the JIRA for testing flows. They're really basic as most of the "fun" flows you can do require that you have a Hadoop cluster up and I didn't want to assume that. If you want some of those for testing, let me know.

mcgilman commented:
@joey I've pulled your changes into develop. Successful build, contrib-check passed, ran a sample flow pulling from one directory and placing into another. Everything looked good. I made a few tweaks to the NOTICE file you provided and rolled up the dependent licenses into the assembly's NOTICE file. Thanks!

joey commented:
Thanks @mcgilman!
